File: LeetCode/85-maximal-rectangle/Solution.java
import java.util.Stack;

class Solution {
  public int maximalRectangle(char[][] matrix) {
    if (matrix.length == 0) {
      return 0;
    }

    int n = matrix.length, m = matrix[0].length;
    int[] heights = new int[m]; // Histogram heights
    int maxArea = 0;

    for (char[] row : matrix) {
      // Update heights array
      for (int j = 0; j < m; j++) {
        heights[j] = (row[j] == '1') ? heights[j] + 1 : 0;
      }
    }

    return maxArea;
  }

  private int largestRectangleArea(int[] heights) {
    Stack<Integer> stack = new Stack<>();
    int maxArea = 0;
    int n = heights.length;

    for (int i = 0; i <= n; i++) {
      int h = (i == n) ? 0 : heights[i];
      while (!stack.isEmpty() && h < heights[stack.peek()]) {
        int height = heights[stack.pop()];
        int width = stack.isEmpty() ? i : i - stack.peek() - 1;
        maxArea = Math.max(maxArea, height * width);
      }
      stack.push(i);
    }
    return maxArea;
  }
}

File: LeetCode/85-maximal-rectangle/solution.ts
/**
 * Implements the monotonic stack optimized approach for the Maximal Rectangle problem.
 *
 * Uses an increasing stack to compute the largest rectangle area efficiently.
 * Processes the matrix row-wise, treating each row as a histogram.
 */
function maximalRectangle(matrix: string[][]): number {
  if (matrix.length === 0) {
    return 0;
  }

  const m = matrix[0].length;
  let maxArea = 0;
  const heights: number[] = new Array(m).fill(0); // Column heights

  // Iterate over each row, treating it as a histogram
  for (const row of matrix) {
    // Update histogram heights
    for (let j = 0; j < m; j++) {
      heights[j] = row[j] === "1" ? heights[j] + 1 : 0;
    }
    // Compute largest rectangle in histogram using monotonic stack
    maxArea = Math.max(maxArea, largestRectangleArea(heights));
  }
  return maxArea;
}

/**
 * Computes the largest rectangle in histogram using a monotonic stack.
 *
 * @param heights Array representing histogram bar heights.
 * @return Maximum rectangular area found within the histogram.
 */
function largestRectangleArea(heights: number[]): number {
  const stack: number[] = []; // Monotonic increasing stack
  let maxArea = 0;
  const n = heights.length;

  for (let i = 0; i <= n; i++) {
    // Assign 0 height to ensure final stack cleanup
    const h = i === n ? 0 : heights[i];

    // Process stored heights while the current bar is shorter
    while (stack.length > 0 && h < heights[stack[stack.length - 1]]) {
      const height = heights[stack.pop()!]; // Height of the popped bar

      // Compute width based on remaining stack indices
      const width = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;

      // Update maxArea with the largest found so far
      maxArea = Math.max(maxArea, height * width);
    }

    // Store current index for future width calculations
    stack.push(i);
  }

  // Return the maximum rectangular area found
  return maxArea;
}
